pkg/ingester/limiting.go

```diff
@@ -0,0 +1,239 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ingester
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/procfs"
+	"github.com/weaveworks/common/httpgrpc"
+)
+
+// This is the closest fitting Prometheus API error code for requests rejected due to limiting.
+const queryLimitingCode = http.StatusServiceUnavailable
+
+func scanCgroupV2() (float64, uint64, error) {
+	// For reference, see https://git.kernel.org/pub/scm/linux/kernel/git/tj/cgroup.git/tree/Documentation/admin-guide/cgroup-v2.rst
+	const memPath = "/sys/fs/cgroup/memory.current"
+	memR, err := readFileNoStat(memPath)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer memR.Close()
+	var memUtil uint64
+	if _, err := fmt.Fscanf(memR, "%d", &memUtil); err != nil {
+		return 0, 0, errors.Wrapf(err, "failed scanning %s", memPath)
+	}
+
+	// For reference, see https://git.kernel.org/pub/scm/linux/kernel/git/tj/cgroup.git/tree/Documentation/admin-guide/cgroup-v2.rst
+	const cpuPath = "/sys/fs/cgroup/cpu.stat"
```

Comment:
Would this approach be portable or would it only work if we're running in a docker container?

Reply:
Do you mean if it would work only in _Docker_ containers, or if it would work outside of containers, period? There is still the fallback on /proc/self, if neither cgroup v2 nor cgroup v1 is present.